arrows/vxl/tests/test_image.cxx

```diff
@@ -181,325 +174,359 @@ populate_vital_image(kwiver::vital::image& img)
+  EXPECT_TRUE( equal_content( img8m, img_loaded ) );
 }
 
+// ----------------------------------------------------------------------------
+TEST_F(image_io_stretch, stretch_on_save)
+{
+  using namespace kwiver;
+
+  vital::image_of<uint8_t> img8{ 200, 300,3 };
+  vital::image_of<uint16_t> img16{ 200, 300, 3 };
+  img16.copy_from( img12 );
+  auto const scale = 255.0 / 4095.0;
+  vital::transform_image( img16, scale_offset<uint16_t>( scale, 0 ) );
+  cast_image( img16, img8 );
+
+  // Test the range stretching at save time
+  auto cs = std::make_shared<simple_image_container>( img12 );
+
+  vital::config_block_sptr config = vital::config_block::empty_config();
+  config->set_value( "auto_stretch", true );
+  config->set_value( "force_byte", true );
+  io.set_configuration( config );
+
+  auto const& image_path = kwiver::testing::temp_file_name( "test-", ".tiff" );
+  io.save( image_path, cs );
+
+  config->set_value( "auto_stretch", false );
+  config->set_value( "force_byte", false );
+  io.set_configuration( config );
+
+  auto const& cl = io.load( image_path );
+  auto const& img_loaded = cl->get_image();
+
+  EXPECT_EQ( img8.pixel_traits(), img_loaded.pixel_traits() );
+  EXPECT_TRUE( equal_content( img8, img_loaded ) );
+
+  EXPECT_EQ( 0, std::remove( image_path.c_str() ) )
+    << "Failed to delete temporary image file.";
+}
 
 namespace {
 
+// ----------------------------------------------------------------------------
 template <typename T>
 void
-run_vil_conversion_tests(const vil_image_view<T>& img, const std::string& type_str)
+run_vil_conversion_tests( vil_image_view<T> const& img )
 {
-  using namespace kwiver::arrows;
-  // convert to a vital image and verify that the properties are correct
-  kwiver::vital::image vimg =  vxl::image_container::vxl_to_vital(img);
-  TEST_EQUAL("VXL image conversion of type "+type_str+" has the correct bit depth",
-             vimg.pixel_traits().num_bytes, sizeof(T));
-  TEST_EQUAL("VXL image conversion of type "+type_str+" has the correct pixel type",
-             vimg.pixel_traits().type, image_pixel_traits_of<T>::static_type);
-  TEST_EQUAL("VXL image conversion of type "+type_str+" has the correct number of planes",
-             vimg.depth(), img.nplanes());
-  TEST_EQUAL("VXL image conversion of type "+type_str+" has the correct width",
-             vimg.height(), img.nj());
-  TEST_EQUAL("VXL image conversion of type "+type_str+" has the correct height",
-             vimg.width(), img.ni());
-  TEST_EQUAL("VXL image conversion of type "+type_str+" has the same memory",
-             vimg.first_pixel(), img.top_left_ptr());
-  bool equal_data = true;
-  for( unsigned int p=0; equal_data && p<img.nplanes(); ++p )
-  {
-    for( unsigned int j=0; equal_data && j<img.nj(); ++j )
+  // Convert to a vital image and verify that the properties are correct
+  image const& vimg = vxl::image_container::vxl_to_vital( img );
+  EXPECT_EQ( sizeof(T), vimg.pixel_traits().num_bytes );
```

Comment:
Can I assume that if one of these tests fails the `EXPECT_EQ` is somehow smart enough to tell me what type `T` is?  In this case it's not enough for the error to say this test failed on line 393.  We need to know which type is being tested.  That is why I was previously passing the type name as a string to print.

Reply:
Alas, no :smile:. However, the *test name* will tell you:

```
  ...
  Test #153: vxl:image_conversion/long.vxl_to_vital
  Test #154: vxl:image_conversion/long.vxl_to_vital_interleaved
  Test #155: vxl:image_conversion/long.vxl_to_vital_cropped
  Test #156: vxl:image_conversion/long.vital_to_vxl
  Test #157: vxl:image_conversion/long.vital_to_vxl_interleaved
  Test #158: vxl:image_conversion/float.vxl_to_vital
  Test #159: vxl:image_conversion/float.vxl_to_vital_interleaved
  Test #160: vxl:image_conversion/float.vxl_to_vital_cropped
  Test #161: vxl:image_conversion/float.vital_to_vxl
  Test #162: vxl:image_conversion/float.vital_to_vxl_interleaved
  ...
```

...or if you run the test binary directly:

```
[----------] 5 tests from image_conversion/6, where TypeParam = long
[ RUN      ] image_conversion/6.vxl_to_vital
[       OK ] image_conversion/6.vxl_to_vital (4 ms)
[ RUN      ] image_conversion/6.vxl_to_vital_interleaved
[       OK ] image_conversion/6.vxl_to_vital_interleaved (4 ms)
[ RUN      ] image_conversion/6.vxl_to_vital_cropped
[       OK ] image_conversion/6.vxl_to_vital_cropped (12 ms)
[ RUN      ] image_conversion/6.vital_to_vxl
[       OK ] image_conversion/6.vital_to_vxl (13 ms)
[ RUN      ] image_conversion/6.vital_to_vxl_interleaved
[       OK ] image_conversion/6.vital_to_vxl_interleaved (16 ms)
[----------] 5 tests from image_conversion/6 (49 ms total)
```

Reply:
...There are several tests like this (not just in this PR; there will be more examples in upcoming PR's) where essentially the same tests were run, differing only by data type, that have been converted to type-parameterized tests. This effectively pushes the "which type?" information up from being printed as part of the failing assertion to the test case name. (Splitting the memory-layout variations into separate test cases is similar.)

This is also why `arrows/vxl/tests/test_image.cxx`, which previously created 4 CTest test cases, now creates 61. (50 of them, as hinted at by the above, are the conversion tests, which test 5 memory layouts for 10 data types: 5×10=50. Previously, this was 1 CTest test case.)